g11n-ws/vip-manager-l10n-agent/src/main/java/com/vmware/l10agent/service/SingleComponentServiceImpl.java

```diff
@@ -213,52 +217,167 @@ private boolean synchkey2VipL10n(RecordModel record,  String key, String srcValu
-	  }
-		return true;
+		ComponentSourceModel model =  getSourceComponentFile(record);
+		if (this.configs.isSyncBatchEnable()){
+			return syncBatchSource(record, model);
+		}else {
+		  return syncSingleSource(record, model);
+		}
+
 	}
-	
+
 	private String postData(String source, String urlStr)  throws VIPHttpException {
 		String response =  HTTPRequester.postData(source, urlStr, "application/json", "POST", null);
 		if(response == null || response.equalsIgnoreCase("")) {
 			throw new VIPHttpException("postFormData error.");
 		}
 		return response;
 	}
+
+	private boolean  syncSingleSource(RecordModel record, ComponentSourceModel model){
+		if (model != null) {
+			for (Entry<String, Object> entry : model.getMessages().entrySet()) {
+				try {
+					boolean result = syncKey2VipL10n(record, entry.getKey(), (String) entry.getValue());
+					if (!result) {
+						result = syncKey2VipI18n(record, entry.getKey(), (String) entry.getValue(), null, null);
+						if (!result) {
+							try {
+								if (record.getStatus() < 5) {
+									record.setStatus(record.getStatus() + 1);
+									TaskSysnQueues.SendComponentTasks.put(record);
+								}
+							} catch (InterruptedException e) {
+
+								Thread.currentThread().interrupt();
+
+								logger.info(e.getMessage(), e);
+							}
+							return result;
+						}
+					}
+				} catch (UnsupportedEncodingException e) {
+
+					logger.error(e.getMessage(), e);
+					return false;
+				}
+			}
+		}
+
+		return true;
+	}
+
+	private  boolean syncBatchSource(RecordModel record, ComponentSourceModel model){
+		if (model != null ){
+			boolean l10nResult = syncBatchKey2VipL10n(record, model);
+			boolean i18nResult = syncBatchKey2VipI18n(record, model);
+
+			if (l10nResult && i18nResult){
+				return true;
+			}else {
+				try {
+					if (record.getStatus() < 5) {
+						record.setStatus(record.getStatus() + 1);
+						TaskSysnQueues.SendComponentTasks.put(record);
+					}
+				} catch (InterruptedException e) {
+					Thread.currentThread().interrupt();
+					logger.error(e.getMessage(), e);
+				}
+				return false;
+			}
+
+		}
+		return true;
+	}
+
+
+	private boolean syncBatchKey2VipI18n(RecordModel record,  ComponentSourceModel model ) {
+
+		if(configs.getVipBasei18nUrl().equalsIgnoreCase(PropertyContantKeys.LOCAL)) {
+			return true;
+		}
+
+		StringBuffer urlStr = new StringBuffer(configs.getVipBasei18nUrl());
+		urlStr.append(PropertyContantKeys.I18N_SOURCE_COLLECT_SET_URL
+				.replace("{" + APIParamName.PRODUCT_NAME + "}", record.getProduct())
+				.replace("{" + APIParamName.VERSION + "}", record.getVersion())
+				.replace("{" + APIParamName.COMPONENT + "}", record.getComponent())
+				.replace("{" + APIParamName.LOCALE + "}", record.getLocale()));
+
+		return prepareBatchDataAndSend(urlStr.toString(), model);
+	}
+
+
+	private boolean syncBatchKey2VipL10n(RecordModel record,  ComponentSourceModel model ){
+		if(configs.getVipBaseL10nUrl().equalsIgnoreCase(PropertyContantKeys.LOCAL)) {
```

Comment:
Is it better to move this into `syncBatchSource`？

Reply:
if we move “if statement” to syncBatchSource，the syncBatchSource method will be complex, and there no different in business. Can you explain your reasons about it

Reply:
These judgement is duplicate in `syncBatchKey2VipI18n` and `syncBatchKey2VipL10n`.
By moving into `syncBatchSource`, it will remove duplicate.

Reply:
syncBatchKey2VipI18n use  configs.getVipBasei18nUrl()  compare， syncBatchKey2VipL10n use  configs.getVipBaseL10nUrl() compare， there no duplicate